diff --git a/src/TokensDateFilterTrait.php b/src/TokensDateFilterTrait.php
index 007c04a..29665c2 100644
--- a/src/TokensDateFilterTrait.php
+++ b/src/TokensDateFilterTrait.php
@@ -29,7 +29,7 @@ trait TokensDateFilterTrait {
   /**
    * {@inheritdoc}
    */
-  public function validateExposed(&$form, FormStateInterface $form_state) {
+  public function validateExposed(&$form, FormStateInterface $form_state):void {
     // Replace tokens to validate date.
     $options = $form_state->getValue('options');
     $value = &$options['value'];

