diff --git a/src/Plugin/GlobalVariable/CurrentDomainName.php b/src/Plugin/GlobalVariable/CurrentDomainName.php
new file mode 100644
index 0000000..c2d76c1
--- /dev/null
+++ b/src/Plugin/GlobalVariable/CurrentDomainName.php
@@ -0,0 +1,33 @@
+<?php
+
+namespace Drupal\mgv\Plugin\GlobalVariable;
+
+use Drupal\mgv\Plugin\GlobalVariable;
+use Drupal\Core\Render\BubbleableMetadata;
+use Drupal\Core\Url;
+use Drupal\domain\DomainInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Component\Utility\Html;
+
+/**
+ * Class CurrentDomainName.
+ *
+ * @package Drupal\mgv\Plugin\GlobalVariable
+ *
+ * @Mgv(
+ *   id = "current_domain_name",
+ * );
+ */
+class CurrentDomainName extends GlobalVariable {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getValue() {
+    $domain = \Drupal::config('domain.settings');
+    $token = \Drupal::token();
+    $domainName = t('[domain:name]');
+    return Html::getClass($token->replace($domainName));
+  }
+
+}
