diff --git a/src/Form/DomainThemeSwitchConfigForm.php b/src/Form/DomainThemeSwitchConfigForm.php
index f1b8d35..9cbc03a 100644
--- a/src/Form/DomainThemeSwitchConfigForm.php
+++ b/src/Form/DomainThemeSwitchConfigForm.php
@@ -2,13 +2,13 @@
 
 namespace Drupal\domain_theme_switch\Form;
 
+use Drupal\Core\Entity\EntityTypeManager;
 use Drupal\Core\Form\ConfigFormBase;
 use Drupal\Core\Form\FormStateInterface;
 use Drupal\Core\Config\ConfigFactoryInterface;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Drupal\Core\Extension\ThemeHandlerInterface;
 use Drupal\Core\Url;
-use Drupal\domain\DomainLoader;
 
 /**
  * Class DomainThemeSwitchConfigForm.
@@ -18,11 +18,11 @@ use Drupal\domain\DomainLoader;
 class DomainThemeSwitchConfigForm extends ConfigFormBase {
 
   /**
-   * Drupal\domain\DomainLoader definition.
+   * The entity type manager.
    *
-   * @var \Drupal\domain\DomainLoader
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
    */
-  protected $domainLoader;
+  protected $entityTypeManager;
 
   /**
    * The theme handler.
@@ -36,15 +36,15 @@ class DomainThemeSwitchConfigForm extends ConfigFormBase {
    *
    * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
    *   The config factory load.
-   * @param \Drupal\domain\DomainLoader $domain_loader
-   *   The domain loader.
+   * @param \Drupal\Core\Entity\EntityTypeManager $entity_type_manager
+   *   The entity type manager.
    * @param \Drupal\Core\Extension\ThemeHandlerInterface $theme_handler
    *   The theme handler.
    */
-  public function __construct(ConfigFactoryInterface $config_factory, DomainLoader $domain_loader, ThemeHandlerInterface $theme_handler
+  public function __construct(ConfigFactoryInterface $config_factory, EntityTypeManager $entity_type_manager, ThemeHandlerInterface $theme_handler
   ) {
     parent::__construct($config_factory);
-    $this->domainLoader = $domain_loader;
+    $this->entityTypeManager = $entity_type_manager;
     $this->themeHandler = $theme_handler;
   }
 
@@ -59,7 +59,7 @@ class DomainThemeSwitchConfigForm extends ConfigFormBase {
    */
   public static function create(ContainerInterface $container) {
     return new static(
-        $container->get('config.factory'), $container->get('domain.loader'), $container->get('theme_handler')
+      $container->get('config.factory'), $container->get('entity_type.manager'), $container->get('theme_handler')
     );
   }
 
@@ -102,7 +102,7 @@ class DomainThemeSwitchConfigForm extends ConfigFormBase {
     $defaultAdminTheme = $this->config('system.theme')->get('admin');
 
     $themeNames = $this->getThemeList();
-    $domains = $this->domainLoader->loadMultipleSorted();
+    $domains = $this->entityTypeManager->getStorage('domain')->loadMultiple();
     foreach ($domains as $domain) {
       $domainId = $domain->id();
       $hostname = $domain->get('name');
@@ -152,7 +152,7 @@ class DomainThemeSwitchConfigForm extends ConfigFormBase {
    */
   public function submitForm(array &$form, FormStateInterface $form_state) {
     parent::submitForm($form, $form_state);
-    $domains = $this->domainLoader->loadMultipleSorted();
+    $domains = $this->entityTypeManager->getStorage('domain')->loadMultiple();
     $config = $this->config('domain_theme_switch.settings');
     foreach ($domains as $domain) {
       $domainId = $domain->id();
