diff --git a/src/EventInstanceCreatorPluginManager.php b/src/EventInstanceCreatorPluginManager.php
index 6bc1a2b..2e3c0bf 100644
--- a/src/EventInstanceCreatorPluginManager.php
+++ b/src/EventInstanceCreatorPluginManager.php
@@ -15,6 +15,13 @@ use Drupal\Core\Plugin\DefaultPluginManager;
  */
 class EventInstanceCreatorPluginManager extends DefaultPluginManager {
 
+  /**
+   * Default plugin ID.
+   *
+   * @var string
+   */
+  protected $defaultPluginId = 'recurring_events_eventinstance_recreator';
+
   /**
    * Constructs a EventInstanceCreatorPluginManager object.
    *
@@ -38,4 +45,14 @@ class EventInstanceCreatorPluginManager extends DefaultPluginManager {
     $this->setCacheBackend($cache_backend, 'eventinstance_creator_info_plugins');
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function createInstance($plugin_id = NULL, array $configuration = []) {
+    // If no plugin ID is provided, use the default plugin ID.
+    if (empty($plugin_id)) {
+      $plugin_id = $this->defaultPluginId;
+    }
+    return parent::createInstance($plugin_id, $configuration);
+  }
 }
