diff --git a/nodeaccess.module b/nodeaccess.module
index 507e3a4..fb4bf71 100644
--- a/nodeaccess.module
+++ b/nodeaccess.module
@@ -76,6 +76,9 @@ function nodeaccess_node_access_records(NodeInterface $node) {
     foreach ($bundle_roles_grants as $role_id => $grant_values) {
       // $grant_values holds values with key grant_view, grant_update and
       // grant_delete.
+      if (!isset($map_rid_gid[$role_id])) {
+        continue;
+      }
       $grants[] = [
         'realm' => 'nodeaccess_role',
         'gid' => $map_rid_gid[$role_id],
@@ -95,6 +98,9 @@ function nodeaccess_node_grants(AccountInterface $account, $op) {
   $grant_ids = [];
 
   foreach ($roles as $role_id) {
+    if (!isset($map_rid_gid[$role_id])) {
+      continue;
+    }
     $grant_ids[] = $map_rid_gid[$role_id];
   }
   $user_id = $account->id();
